### PR TITLE
[SPARK-50116][SQL] Assign appropriate error condition for `_LEGACY_ERROR_TEMP_2067`: `UNSUPPORTED_PARTITION_TRANSFORM`

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to assign proper error condition & sqlstate for `_LEGACY_ERROR_TEMP_2067`: `UNSUPPORTED_PARTITION_TRANSFORM`

### Why are the changes needed?

To improve the error message by assigning proper error condition and SQLSTATE

### Does this PR introduce _any_ user-facing change?

No, only user-facing error message improved

### How was this patch tested?

Updated the existing tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 1]
```diff
@@ -868,8 +868,8 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase with ExecutionE
   def unsupportedPartitionTransformError(
       transform: Transform): SparkUnsupportedOperationException = {
     new SparkUnsupportedOperationException(
-      errorClass = "_LEGACY_ERROR_TEMP_2067",
-      messageParameters = Map("transform" -> transform.toString()))
+      errorClass = "UNSUPPORTED_PARTITION_TRANSFORM",
+      messageParameters = Map("transform" -> toSQLId(transform.toString)))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:872]: Isn't `transform` an expression actually. If so, please, use `toSQLExpr`

File: common/utils/src/main/resources/error/error-conditions.json [hunk 2]
```diff
@@ -5398,6 +5398,12 @@
     },
     "sqlState" : "42902"
   },
+  "UNSUPPORTED_PARTITION_TRANSFORM" : {
+    "message" : [
+      "Unsupported partition transform: <transform>. Ensure that the transform expression is valid."
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-conditions.json:5403]: The error message confuses slightly. Is the transform "unsupported" and "invalid". Can you help users and show which transform is expected: identity, bucket or clusterby.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala (hunk 1 above)
Comment (itholic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:872]: I tried it but unfortunately it says `Transform` is not an `Expression`:

Screenshot 2024-10-29 at 1 54 45 PM

Is there happen to any way converting `Transform` to `Expression`?

File: common/utils/src/main/resources/error/error-conditions.json (hunk 2 above)
Comment (itholic) [reply common/utils/src/main/resources/error/error-conditions.json]: Improved the error message with specific transform names. Thansk!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:875]: ah, I see. `Transform` extends another `Expression` from `org.apache.spark.sql.connector.expressions`. Could you add one more definition of `toSQLExpr` which accept `connector.expressions.Expression`. It should quote in the same way as the existing `toSQLExpr` (using `quoteByDefault`).

Comment (itholic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:875]: Sure will add it. Thanks for the comment!

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:875]: I see. thank you for the investigation.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @itholic.

**Comment by MaxGekk** [comment]
ah, the ticket SPARK-50016 is wrong. @itholic could provide a correct one.

**Comment by itholic** [comment]
Oh, it should be SPARK-50116. Let me update the JIRA

**Pull request closed without merging**